### PR TITLE
feat: add filter strategy with PrefixFilter

### PR DESCRIPTION
**Pull request by zzZIMAWAKE**
Related to: #621

This adds a settings_file_prefix option for toml, json, ini and yaml files which works similarly to envvar_prefix, however it defaults to None and does not filter out any keys if it is not set.

If envs are in use it will do the filtering within each env's variables, an example is shown below

```
[default]
PREFIX_EXTRA = "this will supercede if others are set without prefix"

[development]
PREFIX_HOST = "https://example.com/development"
PREFIX_PORT = "8080"
COMMENT = "this will not show when prefix is used"
EXTRA = "will not show"

[production]
PREFIX_HOST = "https://example.com/production"
PORT = "443"
EXTRA = "will not show"
```

When these are loaded up using the `settings_file_prefix="PREFIX"` option you'll get the following:

```
>>> settings.HOST
'https://example.com/development'
>>> settings.PORT
'8080'
>>> settings.COMMENT
AttributeError: 'Settings' object has no attribute 'COMMENT'
```

When switching to production env:
```
>>> with settings.using_env("production"):
...     settings.HOST
...     settings.EXTRA
...     settings.PORT

'https://example.com/production'
'this will supercede if others are set without prefix'
AttributeError: Attribute PORT was deleted, or belongs to different env
```

I think this behaves as you'd expect, looking forward to some feedback as I had to fumble about a bit to learn how everything fits together so no doubt I've done some less than ideal things

Envless load when environments exist in the file will not work in this way as we'd not be able to detect that `settings.PRODUCTION` is an env to filter the prefix from children. I think this is probably expected if loading envless while envs are included in the file - but if it was possible to check if `ENVIRONMENTS_FOR_DYNACONF` is False *and* detect that the top level key is an environment (`[production]`) then it may be possible, but I could not see an obvious way to detect that `production` is an environment

### CODE REVIEW

**Comment by zzZIMAWAKE** [comment]
I've got to look in to the issue I'm seeing with one of the examples and recursion depth - will take a look at that tomorrow most likely

Will also add the code coverage that I'm missing, happy to receive any feedback now as well if you fancy it :)

**Comment by zzZIMAWAKE** [comment]
Not sure what is happening with the recursive limit being reached for the example: 482_layered_format - it does it for me locally on both master & this branch, have I genuinely broken something?

**Comment by zzZIMAWAKE** [comment]
@rochacbruno This is ready for review, the some parts of the pipeline are taking a while as they are queued but everything seems to pass successfully now.

**Comment by rochacbruno** [comment]
I like this feature, but now that I was able to fully understand its purpose I have some comments:

1. Naming

 `settings_file_prefix` as the name for argument can be confusing, in a first glance it may look like the prefix for a file name instead of variables inside the file.

2. Make it more composable?


Imagine if on the user standing point it can be this:

`dynaconf/strategies/filtering.py`  **new module**
```py

class PrefixFilter:

    def __init__(self, prefix):
        if not isinstance(prefix, str):
            raise TypeError("`SETTINGS_FILE_PREFIX` must be str")
        prefix = "{}_".format(upperfy(prefix))

    def __call__(self, data, settings):
        """Filter incoming data by prefix"""
        prefix = self.prefix
        return {
                upperfy(
                    k[len(prefix):]
                ): v for k, v in data.items()
                if upperfy(k[:len(prefix)]) == prefix
         }

```

on `dynaconf/loaders/base.py` instead of passing `prefix` around as argument we can have the `obj` to hold the `filter_strategy` as an instance of that class.


```py

from dynaconf import Dynaconf
from dynaconf.strategies.filtering import PrefixFilter

settings = Dynaconf(
    settings_file="settings.toml",
    filter_strategy= PrefixFilter("myprefix")
)

```

That `self.filter_strategy` will be a callable that can be used on https://github.com/rochacbruno/dynaconf/pull/625/files#diff-4d2335a3a5b1b1e4cf1f0d9490c21ed91df851ce973b420729a385d372906eb1R167

@zzZIMAWAKE what do you think?

It can accommodate your use case and allow users to be able to customize the filtering with custom classes or any callable that takes `data, settings` arguments.

**Comment by rochacbruno** [comment]
@zzZIMAWAKE I have been thinking in a way to let users to alternatively pass `environments=["prod", "dev", "qa"]` a list of environments explicitly and then "default" and "global" would be automatically part of the environment sets.

It would still accept `True` to load environments arbitrarily but also allow user to specify it. (maybe this is something to be tackled in a different issue/PR)

**Comment by zzZIMAWAKE** [comment]
Would that also be allowed to be passed when `environments=False`? What would be the behaviour in that case? The issue I'm having is that if someone tried to use:

```
from dynaconf import Dynaconf
from dynaconf.strategies.filtering import PrefixFilter

settings = Dynaconf(
    settings_file="settings.toml",
    environments=False
    filter_strategy= PrefixFilter("myprefix")
)
```
with a file that contained:

```
[production]
PREFIX_VAR = TEST
```
It will return blank as it will filter out `production` as it doesn't begin with the prefix so `settings.PRODUCTION.VAR` will not exist as `settings.PRODUCTION` will not even exist. Perhaps this is an edge case that is okay with the new strategy as the user could set up their own filtering strategy.

@rochacbruno Have updated to make use of `filtering_strategy` if you'd like to take a look :)

File: dynaconf/base.py [hunk 1]
```diff
@@ -538,6 +539,10 @@ def from_env(self, env="", keep=False, **kwargs):
             if key not in RENAMED_VARS
         }
 
+        if self.filter_strategy:
+            # Retain the filtering strategy when switching environments
+            new_data["filter_strategy"] = self.filter_strategy
```
Comment (zzZIMAWAKE) [inline dynaconf/base.py:544]: Is this the correct way to retain this when `.using_env` is used? How are other settings retained? (`_validate_only / _validate_exclude` etc, or do they not need to be kept when switching envs?)

**Comment by rochacbruno** [comment]
LGTM  at a forst look @zzZIMAWAKE this is on my list to take deeper a look later today or tomorrow, I will reply here. Thanks again.

File: dynaconf/base.py (hunk 1 above)
Comment (rochacbruno) [reply dynaconf/base.py:544]: validation happens only on first instantiation

**Pull request merged**